### PR TITLE
feat: add stack panel to base component set

### PR DESCRIPTION
**Pull request by scomea**
# Description
The stack panel component enables authors to create a horizontally or vertically scrolling list of child items that can be virtualized (ie. not rendered to the dom) when not within the viewport.  Authors provide the span (height or width depending on orientation) of each item, either a single value when all child items have identical spans, or as an array of values when children have variable spans. Because the stackpanel knows the space to allocate to each item it can create layouts without actually having to render each child which can be a significant performance gain.

The stack panel accepts children in the usual way for html markup.
`<StackPanel>`
`               <button/>`
`                <button/>`
`               <button/> `
`                etc...`
</StackPanel>`
The expectation is that the stackpanel's role is to manage virtualization at React node level and that other parent components, like a data grid or a feed for example, would handle more data driven inputs and handle converting those to nodes to be rendered  in an efficient manner.  

The prototype [data grid component](https://github.com/microsoft/fast-dna/tree/users/scomea/data-grid) takes in data and interacts with the stackpanel to only create nodes only for the rows that are in the region that is being rendered.  Browsing a large dataset in this scenario could end up with a stackpanel with children that look something like this:

`<StackPanel itemSpan={[500000, 100, ...more spans...100, 300000]}>`
`                <div>Filler child that represents first 5000 rows</div>`
`                <dataRow>Row index 5001</dataRow>`
`               ...more data rows...`
 `             <dataRow>Row index 6000</dataRow>`
 `           <div>Filler child that represents rows after 6000</div>`
`</StackPanel>`

As the user scrolls the data grid tracks the scroll position and updates the children as required.  The first and last child merely exist to take up virtual space and their spans change as the component is scrolled and which data rows are actually passed to the data grid changes accordingly.

## Motivation & context
https://github.com/microsoft/fast-dna/issues/2501

Consider a stackpanel with 200 child items rendered with virtualization turned off which takes about 5 seconds of churn on the main thread when first rendered:

200NoVirt

Turning on virtualization reduces that time to less than a second:

200wVirt

Further, upping the item count from 200 to 10000 does not have a significant impact:

10kwVirt

## Issue type checklist
- [ ] **Chore**: A change that does not impact distributed packages.
- [ ] **Bug fix**: A change that fixes an issue, link to the issue above.
- [x] **New feature**: A change that adds functionality.

**Is this a breaking change?**
- [ ] This change causes current functionality to break.

## Process & policy checklist
- [x] I have added tests for my changes.
- [x] I have tested my changes.
- [x] I have updated the project documentation to reflect my changes.
- [x] I have read the [CONTRIBUTING](https://github.com/Microsoft/fast-dna/blob/master/CONTRIBUTING.md) documentation and followed the [standards](https://www.fast.design/docs/en/contributing/standards) for this project.

### CODE REVIEW

File: packages/fast-components-react-base/src/stack-panel/stack-panel.tsx [hunk 1]
```diff
@@ -0,0 +1,738 @@
...
+import {
+    StackPanelHandledProps,
+    stackPanelOrientation,
+    StackPanelProps,
+    StackPanelUnhandledProps,
+} from "./stack-panel.props";
+import { isFunction } from "util";
+
+export interface StackPanelState {
+    disabled: boolean;
```
Comment (nicholasrice) [inline packages/fast-components-react-base/src/stack-panel/stack-panel.tsx]: I don't see disabled used anywhere - is it needed?

File: packages/fast-components-react-base/src/stack-panel/stack-panel.tsx [hunk 2]
```diff
@@ -0,0 +1,738 @@
...
+    isScrolling: boolean;
+    renderedRangeStartIndex: number;
+    renderedRangeEndIndex: number;
+    direction: Direction;
+}
+
+interface ItemPosition {
```
Comment (nicholasrice) [inline packages/fast-components-react-base/src/stack-panel/stack-panel.tsx]: What do the values in `ItemPosition` represent? Are these pixel offset values? Is some of this information derived?

File: packages/fast-components-react-base/src/stack-panel/stack-panel.tsx [hunk 3]
```diff
@@ -0,0 +1,738 @@
...
+     */
+    public render(): React.ReactElement<HTMLDivElement> {
+        const {
+            stackPanel_itemContainer,
+        }: StackPanelClassNameContract = this.props.managedClasses;
+
+        return (
+            <div
+                {...this.unhandledProps()}
+                tabIndex={-1}
```
Comment (nicholasrice) [inline packages/fast-components-react-base/src/stack-panel/stack-panel.tsx]: Why do both of these elements need tabIndex?

File: packages/fast-components-react-base/src/stack-panel/stack-panel.tsx [hunk 4]
```diff
@@ -0,0 +1,738 @@
...
+                      left: toPx(this.itemPositions[index].start),
+                  }
+                : {
+                      height: toPx(this.itemPositions[index].span),
+                      top: toPx(this.itemPositions[index].start),
+                  }),
+        };
+
+        if (this.isClonableElement(item)) {
+            const styleProps: React.CSSProperties = get(item.props, "style");
```
Comment (nicholasrice) [inline packages/fast-components-react-base/src/stack-panel/stack-panel.tsx]: are we cloning *every* child?

File: packages/fast-components-react-base/src/stack-panel/stack-panel.tsx (hunk 1 above)
Comment (scomea) [reply packages/fast-components-react-base/src/stack-panel/stack-panel.tsx]: Let's pull it until we figure out we need it.

File: packages/fast-components-react-base/src/stack-panel/stack-panel.tsx (hunk 2 above)
Comment (scomea) [reply packages/fast-components-react-base/src/stack-panel/stack-panel.tsx]: Added comment "Used to store the pixel coordinates and span of items along the stackpanel's axis".  Basically it's used to create a lookup table of the coordinates and spans of the children on mount and update (see "updateItemPositions()") so we don't have to recalculate on scroll events.

File: packages/fast-components-react-base/src/stack-panel/stack-panel.tsx (hunk 3 above)
Comment (scomea) [reply packages/fast-components-react-base/src/stack-panel/stack-panel.tsx]: removed, may not be necessary

File: packages/fast-components-react-base/src/stack-panel/stack-panel.tsx (hunk 4 above)
Comment (scomea) [reply packages/fast-components-react-base/src/stack-panel/stack-panel.tsx]: Only the ones we actually render - those in the viewport and configurable buffer region, and those in the "never virtualize" list.  Of course if we turn off virtualization we clone them all

Comment (scomea) [reply packages/fast-components-react-base/src/stack-panel/stack-panel.tsx]: Might be an optimization possible *if* we can safely cache the clones between updates.

File: packages/fast-components-class-name-contracts-base/src/stack-panel/index.ts [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+export interface StackPanelClassNameContract {
+    /**
+     * The root of the stackpanel component
+     */
+    stackPanel?: string;
+
+    /**
+     * the container that contains the stack panel items
+     */
+    stackPanel_itemContainer?: string;
```
Comment (janechu) [inline packages/fast-components-class-name-contracts-base/src/stack-panel/index.ts]: I would prefer to just say items, so `stackPanel_items`.

File: packages/fast-components-class-name-contracts-base/src/stack-panel/index.ts [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+    /**
+     *
+     */
+    stackPanel__isScrolling?: string;
```
Comment (janechu) [inline packages/fast-components-class-name-contracts-base/src/stack-panel/index.ts]: just as we would say `stackPanel__active`, and not `stackPanel__isActive` I prefer `stackPanel__scrolling`.

Also, think you need a comment above this.

File: packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.spec.tsx [hunk 7]
```diff
@@ -347,6 +347,24 @@ describe("horizontal overflow", (): void => {
             0.0004
         );
     });
+    test("should ease the animation correctly when at animation start time", () => {
```
Comment (janechu) [inline packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.spec.tsx]: I know it's rife through this file, but I prefer to leave out the words "correct" and "correctly" rather "should do x" would suffice since all actions should be performed correctly. In this case I think it might be better to re-title the test "should use an ease animation at the animation start time".

File: packages/fast-components-react-base/src/index.ts [hunk 8]
```diff
@@ -86,6 +86,10 @@ import SliderTrackItem from "./slider-track-item";
 export { SliderTrackItem };
 export * from "./slider-track-item";
 
+import StackPanel from "./stack-panel";
+export { StackPanel };
+export * from "./stack-panel";
```
Comment (janechu) [inline packages/fast-components-react-base/src/index.ts]: I'm not sure this should be available from the index location due to the note on this still being in initial development.

File: packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+export interface ItemSpanOverride {
+    [key: number]: number;
+}
+
+export interface StackPanelHandledProps extends StackPanelManagedClasses {
+    /**
+     * Whether items outside the viewport should be rendered or not
+     * default is true
+     */
+    enableVirtualization?: boolean;
```
Comment (janechu) [inline packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts]: I think this could just be `virtualize`.

File: packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+
+    /**
+     *  the indexes of children that should always be rendered
```
Comment (janechu) [inline packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts]: Can you provide a use case?

File: packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+     */
+    neverVirtualizeIndexes?: number[];
+
+    /**
+     *  The default span of each item (width in horizontal orientation, height in vertical)
```
Comment (janechu) [inline packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts]: Not sure that "span" gets across the dimensionality of the items, also not sure if you shouldn't just pass the width and height separately and let the user input them, perhaps `itemWidth` and `itemHeight`.

File: packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+     */
+    defaultItemSpan?: number;
+
+    /**
+     * An object whose entries specifies the height of items that differ from the default height.
+     * Format is {itemIndex: heightValue}
+     */
+    itemSpanOverrides?: ItemSpanOverride;
```
Comment (janechu) [inline packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts]: Is this necessary in the first iteration of this component? I think you can specify an override strategy and have it be a feature.

File: packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+
+    /**
+     * How many items to preload before and after the range defined by viewportMaxHeight
+     */
+    preloadBufferLength?: number;
```
Comment (janechu) [inline packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts]: Should this be settable for now? I would think that either there would be some kind of calculation we are doing to determine how much of a buffer we should have or we allow a function to be passed that would provide a calculation based on some parameters like item size and viewport size.

File: packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+
+    /**
+     * The orientation of the stackpanel
+     */
+    orientation?: stackPanelOrientation;
```
Comment (janechu) [inline packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts]: I think we have orientation in the `fast-web-utilities/src/aria.ts` you could use

File: packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+
+    /**
+     * The children of the panel
+     */
+    children?: React.ReactNode;
```
Comment (janechu) [inline packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts]: I think children doesn't need to be specified

File: packages/fast-components-react-base/src/stack-panel/stack-panel.schema.ts [hunk 16]
```diff
@@ -0,0 +1,58 @@
+export default {
+    $schema: "http://json-schema.org/schema#",
+    title: " stack panel",
```
Comment (janechu) [inline packages/fast-components-react-base/src/stack-panel/stack-panel.schema.ts]: nit: space in from of title, would capitalize the "S".

File: packages/fast-components-react-base/src/stack-panel/stack-panel.tsx [hunk 17]
```diff
@@ -0,0 +1,738 @@
...
+        }
+
+        let renderEndIndex: number = visibleRangeEndIndex + bufferLength;
+        if (renderEndIndex > lastIndex) {
+            renderEndIndex = lastIndex;
+        }
+        this.setState({
+            renderedRangeStartIndex: renderStartIndex,
+            renderedRangeEndIndex: renderEndIndex,
+            isScrolling: this.itemContainerSpan > this.viewPortSpan ? true : false,
```
Comment (janechu) [inline packages/fast-components-react-base/src/stack-panel/stack-panel.tsx]: can just be `this.itemContainerSpan > this.viewPortSpan`, don't need a ternary for true/false.

File: packages/fast-components-react-base/src/stack-panel/stack-panel.tsx [hunk 18]
```diff
@@ -0,0 +1,738 @@
...
+     * Builds a list of positioning data for all items
+     */
+    private updateItemPositions = (): void => {
+        const itemPositions: ItemPosition[] = [];
+        let currentItemTop: number = 0;
+        const childrenAsArray: React.ReactNode[] = React.Children.toArray(
+            this.props.children
+        );
+        const itemCount: number = childrenAsArray.length;
+        const keys: string[] = Object.keys(this.props.itemSpanOverrides);
```
Comment (janechu) [inline packages/fast-components-react-base/src/stack-panel/stack-panel.tsx]: What's the benefit of the itemSpanOverrides not being just an array?

File: packages/fast-components-react-base/src/stack-panel/stack-panel.tsx (hunk 17 above)
Comment (janechu) [inline packages/fast-components-react-base/src/stack-panel/stack-panel.tsx]: Seems to me if the itemContainerSpan is larger than the viewPortSpan, then that implies it is scrollable, not the isScrolling is happening.

File: packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+
+    /**
+     * When the component scrolls focused content into view how many pixels should the next item "peek" into view
+     */
+    nextItemPeek?: number;
+
+    /**
+     * Whether to enable smooth scrolling
+     */
+    enableSmoothScrolling?: boolean;
```
Comment (janechu) [inline packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts]: I wonder if we can't just provide a prop that is an animation function that can be passed, so instead of having a smooth scrolling function, we have a secondary smooth scrolling function they can pass in to a `scrollAnimation` prop

File: packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+
+    /**
+     * Callback for when scroll values change
+     */
+    onScrollChange?: (newScrollValue: number, scrollMaxValue: number) => void;
+
+    /**
+     * When the component scrolls focused content into view how many pixels should the next item "peek" into view
+     */
+    nextItemPeek?: number;
```
Comment (janechu) [inline packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts]: should we instead of offering a peek, just make this controlled with regards to the scrolled distance?

File: packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts (hunk 14 above)
Comment (janechu) [inline packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts]: Should we even have orientation? Can virtualization happen in both horizontal and vertical axis? Take excel for example.

File: packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts (hunk 10 above)
Comment (scomea) [reply packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts]: The scenario you can see working in the stories is the case where we always render the first and last item in the list so that tabbing to the component brings one of those into focus.  The other use case is if someone wanted to implement something that resembled lazy loading where some items don't unload after they have been loaded.

File: packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts (hunk 12 above)
Comment (scomea) [reply packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts]: I want to bring over a datagrid component to use this and it would require variable spans.

File: packages/fast-components-react-base/src/stack-panel/stack-panel.tsx (hunk 18 above)
Comment (scomea) [reply packages/fast-components-react-base/src/stack-panel/stack-panel.tsx]: It needs to be keyed as it only specifies the items that vary from the default span (ie. it does not need to include an entry for each item). I thought about using a Map but didn't think that seemed better.

File: packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts (hunk 19 above)
Comment (scomea) [reply packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts]: I've thought about having a function to calculate the animation curve but I'm not sure we should do that instead of a Boolean.  I'm thinking we want the scrolling on by default but if all we have is a function prop then how do you turn it off?

File: packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts (hunk 14 above)
Comment (scomea) [reply packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts]: Yes, IMO we should have orientation.  A horizontal carousel could totally benefit from virtualization, for example,

File: packages/fast-components-react-base/src/stack-panel/README.md [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+
+Authors should be aware that this component writes to the width/left or top/height css style attributes of children on render depending on the orientation of the stack panel.
+
+### Props
+
+ enableVirtualization:  Whether child elements are virtualized or not. Defaults to true.
+
+ neverVirtualizeIndexes: Authors can provide an array of indexes which flag child indexes that should not be virtualized.  For example providing [0, 10] to this prop would mean that child items at indexes 0 and 10 would be rendered.
+
+ defaultItemSpan:  The default span of the child items in pixels, this can be overriden on an item by item basis using the "itemSpanOverrides" prop. Default is 100.
```
[the remark this answers is not in the document]
Comment (scomea) [reply packages/fast-components-react-base/src/stack-panel/README.md]: Yes, "span" is meant to refer to height or width based on orientation.  Happy to change the term if someone has a better idea,  "size" works for me.

[the remark this answers is not in the document]
Comment (nicholasrice) [reply packages/fast-components-react-base/src/stack-panel/README.md]: I had this thought also, but then looked up at it seemed accurate to me - in particular the example:

> the full extent of something from end to end; the amount of space that something covers.
"a warehouse with a clear span of 28 feet"

https://www.google.com/search?sxsrf=ACYBGNR6PRYcr1tTn75z_MHjdyiuCCSBOA%3A1579799345198&ei=MdMpXunaC4HN-gTT14_IDQ&q=span+definition&oq=span+definition&gs_l=psy-ab.3..35i39l2j0l8.2918.7009..7253...1.2..0.62.578.12......0....1..gws-wiz.......0i71j0i67j0i131j0i10j0i22i30j0i22i10i30.5vqFrfo9VXU&ved=0ahUKEwipt6WWm5rnAhWBpp4KHdPrA9kQ4dUDCAs&uact=5

"size" seems good to me also, and avoids that conflation

File: packages/fast-components-react-base/src/stack-panel/stack-panel.tsx (hunk 4 above)
Comment (nicholasrice) [reply packages/fast-components-react-base/src/stack-panel/stack-panel.tsx]: should we expect a render function instead of children, so that we can pass the required data as an arg to that cb, instead of cloning children?

File: packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts (hunk 14 above)
Comment (janechu) [reply packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts]: I'm in agreement, I meant what if it was virtualizing both horizontally and vertically.

File: packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts (hunk 19 above)
Comment (janechu) [reply packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts]: Well, this is in base correct? Animation is something I would consider as part of styling, so if anything I would offer a prop to inject the animation function, any implementation in a styled lib such as `fast-component-styles-msft` would then have the exported animation. Maybe @bheston or @nicholasrice has an opinion.

File: packages/fast-components-react-base/src/stack-panel/stack-panel.tsx (hunk 18 above)
Comment (janechu) [reply packages/fast-components-react-base/src/stack-panel/stack-panel.tsx]: What I mean is, why do:
```
{
    "0": "foo",
    "1": "bar"
}
```

and not just:
```
[
    "foo",
    "bar"
]
```

File: packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts (hunk 11 above)
Comment (scomea) [reply packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts]: The only span/size we are concerned with is the along the scrolling axis of the panel, so it doesn't seem ideal to take both height and width.  I think that would be confusing as we'd just be throwing away the value that isn't on the axis we scroll on.

File: packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts (hunk 13 above)
Comment (scomea) [reply packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts]: I think it's useful and straightforward.  It does not mean that we can't have some calculation that figures out the buffer based on an itemheight/containerHeight calculation.

File: packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts (hunk 14 above)
Comment (scomea) [reply packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts]: I was not intending to support virtualization on both axis simultaneously in this component.  The scenarios this is intended for initially is a feed and datagrid scenario which does not require more than one axis.

File: packages/fast-components-class-name-contracts-base/src/stack-panel/index.ts (hunk 5 above)
Comment (scomea) [reply packages/fast-components-class-name-contracts-base/src/stack-panel/index.ts]: sure

File: packages/fast-components-class-name-contracts-base/src/stack-panel/index.ts (hunk 6 above)
Comment (scomea) [reply packages/fast-components-class-name-contracts-base/src/stack-panel/index.ts]: done

File: packages/fast-components-react-base/src/index.ts (hunk 8 above)
Comment (scomea) [reply packages/fast-components-react-base/src/index.ts]: good idea

File: packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts (hunk 9 above)
Comment (scomea) [reply packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts]: done

File: packages/fast-components-react-base/src/stack-panel/stack-panel.tsx (hunk 4 above)
Comment (scomea) [reply packages/fast-components-react-base/src/stack-panel/stack-panel.tsx]: I've thought about that a fair bit and have the idea that the base stackpanel component should accept children directly so that an author could populate it directly with markup, I think that's valuable.  Components derived from this one, like say a datagrid, would be the place where data inputs are converted to children with render functions and could use callback events that pass the rendered range to update children as the component is interacted with.  This could also support a pagination strategy where the derived component could pass a single child that represents a range of out of view children as a "page" but maintains the height.  Working on a proof of concept for this.

File: packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts (hunk 19 above)
Comment (bheston) [reply packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts]: As we chatted, I like the idea of not baking the animation into the component, though I understand the particular list / tabbing issue you're solving. If there's a good way to encapsulate this and allow it to be easily updated I think that makes sense.

File: packages/fast-components-react-base/src/stack-panel/stack-panel.schema.ts [hunk 22]
```diff
@@ -0,0 +1,58 @@
...
+    description: "A  stack panel component's schema definition.",
```
Comment (eljefe223) [inline packages/fast-components-react-base/src/stack-panel/stack-panel.schema.ts]: extra space

File: packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts (hunk 14 above)
Comment (scomea) [reply packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts]: done

File: packages/fast-components-react-base/src/stack-panel/stack-panel.tsx [hunk 23]
```diff
@@ -0,0 +1,738 @@
...
+            <div onFocusCapture={this.onItemFocus} style={newStyleProps}>
+                {item}
+            </div>
+        );
+    };
+
+    /**
+     * Builds a list of positioning data for all items
+     */
+    private updateItemPositions = (): void => {
```
[the remark this answers is not in the document]
Comment (scomea) [reply packages/fast-components-react-base/src/stack-panel/stack-panel.tsx]: The developer using the stack panel must be able to provide correct dimensions for the children that take into account whatever css is applied to them.  We don't measure dom elements in any way.

File: packages/fast-components-react-base/src/stack-panel/stack-panel.schema.ts (hunk 16 above)
Comment (scomea) [reply packages/fast-components-react-base/src/stack-panel/stack-panel.schema.ts]: done

File: packages/fast-components-react-base/src/stack-panel/stack-panel.tsx (hunk 18 above)
Comment (scomea) [reply packages/fast-components-react-base/src/stack-panel/stack-panel.tsx]: Because if we do a straight array we need to populate a value for every item, with keys we can limit to just those items that deviate from the default height.  Think a data grid that has a large number of rows but only the one the user is operating on is expanded.  With keys we can do:
 {
    "375": 200,
}

vs. 
[
    …, to the beginning
    100,
    100,
    200,
    100,
    100,
    … to the end
]

File: packages/fast-components-react-base/src/stack-panel/stack-panel.tsx (hunk 17 above)
Comment (scomea) [reply packages/fast-components-react-base/src/stack-panel/stack-panel.tsx]: yep

File: packages/fast-components-react-base/src/stack-panel/README.md (hunk 21 above)
[the remark this answers is not in the document]
Comment (scomea) [reply packages/fast-components-react-base/src/stack-panel/README.md]: Folks seem to be ok with "span" after some discussion.  Please reactivate if further discussion needed.

File: packages/fast-components-react-base/src/stack-panel/stack-panel.schema.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (scomea) [reply packages/fast-components-react-base/src/stack-panel/stack-panel.schema.ts]: yep

File: packages/fast-components-react-base/src/stack-panel/stack-panel.tsx [hunk 24]
```diff
@@ -0,0 +1,738 @@
...
+        if (!isNil(this.rootElement.current)) {
+            if (((window as unknown) as WindowWithResizeObserver).ResizeObserver) {
+                this.resizeDetector = new ((window as unknown) as WindowWithResizeObserver).ResizeObserver(
+                    this.handleResize
+                );
+                this.resizeDetector.observe(this.rootElement.current);
+            }
+        }
+
+        this.viewPortSpan = this.getViewportSpan();
```
[the remark this answers is not in the document]
Comment (scomea) [reply packages/fast-components-react-base/src/stack-panel/stack-panel.tsx]: yep

File: packages/fast-components-react-base/src/stack-panel/stack-panel.schema.ts (hunk 22 above)
Comment (scomea) [reply packages/fast-components-react-base/src/stack-panel/stack-panel.schema.ts]: zapped

File: packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts [hunk 25]
```diff
@@ -0,0 +1,84 @@
...
+import {
+    ManagedClasses,
+    StackPanelClassNameContract,
+} from "@microsoft/fast-components-class-name-contracts-base";
+
+export interface StackPanelManagedClasses
+    extends ManagedClasses<StackPanelClassNameContract> {}
+export interface StackPanelUnhandledProps extends React.HTMLAttributes<HTMLDivElement> {}
+
+export enum stackPanelOrientation {
```
[the remark this answers is not in the document]
Comment (scomea) [reply packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts]: enum no longer exists

File: packages/fast-components-react-base/src/stack-panel/stack-panel.tsx (hunk 4 above)
Comment (scomea) [reply packages/fast-components-react-base/src/stack-panel/stack-panel.tsx]: See the concept working in the data-grid working branch 
https://github.com/microsoft/fast-dna/tree/users/scomea/data-grid

File: packages/fast-components-react-base/src/stack-panel/stack-panel.tsx (hunk 18 above)
Comment (janechu) [reply packages/fast-components-react-base/src/stack-panel/stack-panel.tsx]: You do not need to populate a real value for every item, if you did `var test = new Array(100)` you would get an array of 100 empty items.

**Comment by janechu** [comment]
I'd like to backtrack on this PR a little, I think this is too complex a task to not have an associated issue filed before this feature can be put in.

I think we need to update our contributor documentation, but what I would suggest is closing this issue and filing a feature request that includes:

- abstract: high level description of the component and what it should do
- background: any background information needed to understand the ask
- proposed implementation: how will the component be implemented
- additional information: any extra information that should be taken into consideration (optional)

File: packages/fast-components-react-base/src/stack-panel/stack-panel.tsx (hunk 18 above)
Comment (scomea) [reply packages/fast-components-react-base/src/stack-panel/stack-panel.tsx]: Replaced "defaultItemSpan" prop with simply "itemSpan" that takes either a single number for uniform sized items or a flat array (number[]) for variable sized items.  Seems more intuitive.

Comment (scomea) [reply packages/fast-components-react-base/src/stack-panel/stack-panel.tsx]: Switched to a number[] array for simplicity.

File: packages/fast-components-react-base/src/stack-panel/README.md [hunk 26]
```diff
@@ -0,0 +1,40 @@
...
+
+preloadBufferLength: How many out of view child items to render on either side of the viewport. Default is 1.
+
+orientation: Whether the child elements stack vertically or horizontally.  Default is vertical.
+
+onScrollChange:  Callback for when the component scrolls.
+
+nextItemPeek:  When the component scrolls focused content into view how many pixels should the next item "peek" into view.  Note that the component will clip the peek value applied based on available viewport size.  Default is 50.
+
+enableSmoothScrolling:  Whether smooth scrolling is enabled or not.  When disabled focused items will be brought into view by the browser (ie. not animated).  Default is true.
```
[the remark this answers is not in the document]
Comment (scomea) [reply packages/fast-components-react-base/src/stack-panel/README.md]: removed animation code as I found a css based solution

File: packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts (hunk 20 above)
Comment (scomea) [reply packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts]: removed animation code as I found a css based solution

File: packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts (hunk 19 above)
Comment (scomea) [reply packages/fast-components-react-base/src/stack-panel/stack-panel.props.ts]: removed animation code as I found a css based solution
